gapis/server/export_replay.go

```diff
@@ -0,0 +1,156 @@
+	if d == nil {
+		return log.Errf(ctx, nil, "Unable to produce replay on unknown device.")
+	}
+
+	cap, err := capture.ResolveFromPath(ctx, c)
+	intent := replay.Intent{d, c}
+
+	var queries []func(mgr replay.Manager) error
+	switch {
+	case opts.Report != nil && len(opts.GetFramebufferAttachmentRequests) > 0:
+		return log.Errf(ctx, nil, "Report and Framebuffer requests are not compatible.")
+	case opts.GetFramebufferAttachmentRequests != nil:
+		r := &path.ResolveConfig{ReplayDevice: d}
+		changes, err := resolve.FramebufferChanges(ctx, c, r)
+		if err != nil {
+			return err
+		}
+
+		for _, req := range opts.GetFramebufferAttachmentRequests {
+			req := req
+			fbInfo, err := changes.Get(ctx, req.After, req.Attachment)
+			if err != nil {
+				return err
+			}
+
+			for _, a := range cap.APIs {
+				a, ok := a.(replay.QueryFramebufferAttachment)
+				if !ok {
+					continue
+				}
+				queries = append(queries, func(mgr replay.Manager) error {
+					_, err := a.QueryFramebufferAttachment(
+						ctx,                   // context.Context
+						intent,                // Intent
+						mgr,                   // Manager
+						req.After.Indices,     // after []uint64
+						fbInfo.Width,          // width uint32
+						fbInfo.Height,         // height uint32
+						req.Attachment,        // api.FramebufferAttachment
+						fbInfo.Index,          // uint32
+						req.Settings.DrawMode, // service.DrawMode
+						true,  // disableReplayOptimization bool
+						false, // displayToSurface bool
+						nil,   // hints *service.UsageHints
+					)
+					return err
+				})
+			}
+		}
+	case opts.Report != nil:
+		// TODO(hysw): Add a simple replay request that output commands as
+		// captured. For now, if there are no frame query, force an issue query.
+		// Since otherwise the trace will not be generated.
+		fallthrough
+	default:
+		for _, a := range cap.APIs {
+			a, ok := a.(replay.QueryIssues)
+			if !ok {
+				continue
+			}
+			queries = append(queries, func(mgr replay.Manager) error {
+				_, err := a.QueryIssues(ctx, intent, mgr, false, nil)
+				return err
+			})
+		}
+	}
+
+	exporter := replay.NewExporter()
+
+	errs := make(chan error, len(queries))
+	for _, q := range queries {
+		q := q
+		go func() {
+			errs <- q(exporter)
+		}()
+	}
+
+	payload, err := exporter.Export(ctx, len(queries))
+	if err != nil {
+		return err
+	}
+
+	for range queries {
+		err := <-errs
+		if err != nil {
+			// TODO(hysw): Ignore error for now, since empty postback will upset some
+			// of the postback handlers.
+		}
+	}
+
+	err = os.MkdirAll(out, os.ModePerm)
+	if err != nil {
+		return log.Errf(ctx, err, "Failed to create output directory: %v", out)
+	}
+
+	payloadBytes, err := proto.Marshal(payload)
+	if err != nil {
+		return log.Errf(ctx, err, "Failed to serialize replay payload.")
+	}
+	err = ioutil.WriteFile(gopath.Join(out, "payload.bin"), payloadBytes, 0644)
```

Comment:
Would be nicer if the server handed back the data to the client so that we don't have to worry about  all the `enableLocalFiles` stuff.

Reply:
There's multiple files to be written: payload.bin, archive.{index,data} so it is not that easy to stream the data back. The code produce those the archive is in C land at core/cc/archive.*.

Reply:
Is all the gapir changes just so you can use the archive C++ library?
It does strike me as odd that we're writing files from two separate processes here.

Reply:
Hold on, I'm talking nonsense. Let me just check out this PR and figure out what the hell I'm talking about.

Reply:
Right, ignore the craziness I said above.

So we're writing out 3 files - `payload.bin` (this line) and `resources.index`, `resource.data` (below).
I don't really get why this couldn't just be put into a filename-data map which is returned, and handed back over gRPC.

However, I'm not suggesting making these changes right now - I just want to know if it is possible.

Reply:
In that case, we either need to build all the files in memory (resource.data can be many GBs) or use a streaming rpc. And either way we need to duplicate archive.* code in Go somehow. So it is possible, however I don't see a compelling reason to build it and maintain it.